odxtools/progcode.py

```diff
@@ -52,4 +56,9 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
         pass
 
     def _resolve_snrefs(self, diag_layer: "DiagLayer") -> None:
-        pass
+        db = diag_layer.database
+
+        self._code = odxrequire(
+            db.auxiliary_files.get(self.code_file),
```

Comment:
This will kill performance for flashing PDX files, since they can contain large binary data, better to provide a lazy API that only extracts the binary bytes when needed.

Reply:
it was my assumption that python uses a reference here? (a copy would be `bytes(db.auxiliary_files.get(self.code_file))`?) Maybe I'm wrong, though. Can you test or do you know a good reference to the documentation?

Reply:
okay, I ran two tests. The first was
```python
import tracemalloc

tracemalloc.start()

x = chr(0)*1000*1000
db = []
for i in range(999):
    #x = chr(i%256)*1000*1000
    db.append(x)

print(tracemalloc.get_traced_memory())

tracemalloc.stop()
```

and the second
```python
import tracemalloc

tracemalloc.start()

x = chr(0)*1000*1000
db = []
for i in range(999):
    #x = chr(i%256)*1000*1000
    db.append(x)

print(tracemalloc.get_traced_memory())

tracemalloc.stop()
```

for the first test, the memory consumption was basically identical to not having the `db` object (8832 bytes overhead for the list with 999 items), the second used almost exactly 999 times more memory than the first script (999.07 MB vs 1.01 MB). What I considered interesting is when the first script was modified to copy the items: this resulted in exactly the same memory consumption, so python seems to use a "copy on write" mechanism for strings...

Reply:
Comment was on the wrong line, I am worried about memory consumed by the `aux_file_data = open(aux_file_name, "rb").read()` operation

Reply:
I think we can assume that odxtools are used on desktop class machines, i.e. there should be a few gigabytes of RAM available? What's IMO more worrying is the size of the uncompressed ODX XML files: I've git a PDX which contains an `*.odx-d` file exhibiting a size of > 650 MB, and parsing this to an `ElementTree` requires quite a bit of additional RAM, i.e. loading this file on a 4GB RaspberryPi is "fun"...

Reply:
PDX file that ship ECU firmware can be even bigger, multiple gigabytes of binary that basically contain a whole OS and its applications, those kind of auxiliary_files should only be loaded when explicitly requested, otherwise you run out of RAM fast.

Reply:
I see. fixed: https://github.com/mercedes-benz/odxtools/pull/310/commits/fe5423a7f2033a3fd9f46ab1c0d98e29cb744ae1